### PR TITLE
Cvs 91634 metric family

### PR DESCRIPTION
**Pull request by rasapala**
Enabling metrics with family names and changing bool handling to sets.

### CODE REVIEW

File: src/metric_config.hpp [hunk 1]
```diff
@@ -30,73 +32,64 @@ class MetricConfig {
...
+            {"ovms_requests_success_grpc_predict",
+                "ovms_requests_success_grpc_getmodelmetadata",
+                "ovms_requests_success_grpc_getmodelstatus",
+                "ovms_requests_success_grpc_modelinfer",
+                "ovms_requests_success_grpc_modelmetadata",
+                "ovms_requests_success_grpc_modelready",
+                "ovms_requests_success_rest_modelinfer",
+                "ovms_requests_success_rest_predict",
+                "ovms_requests_success_rest_get_modelmetadata",
+                "ovms_requests_success_rest_get_modelstatus",
```
Comment (dkalinowski) [inline src/metric_config.hpp:54]: same with `get_modelstatus`

File: src/model_metric_reporter.cpp [hunk 2]
```diff
@@ -42,171 +42,176 @@ ServableMetricReporter::ServableMetricReporter(const MetricConfig* metricConfig,
         this->buckets.emplace_back(floor(BUCKET_MULTIPLIER * pow(BUCKET_POWER_BASE, i)));
     }
 
-    auto family = registry->createFamily<MetricCounter>("ovms_requests_success",
+    auto familyName = "ovms_requests_success";
```
Comment (dkalinowski) [inline src/model_metric_reporter.cpp:45]:
```suggestion
    std::string familyName = "ovms_requests_success";
```
Using `auto` in this context might be std::string or const char*. Probably const char. Why wouldnt we define what type we want?

File: src/metric_config.hpp [hunk 3]
```diff
@@ -30,73 +32,64 @@ class MetricConfig {
...
+                "ovms_requests_success_rest_modelmetadata",
+                "ovms_requests_success_rest_modelready"}},
+        {"ovms_requests_fail",
+            {"ovms_requests_fail_grpc_predict",
+                "ovms_requests_fail_grpc_getmodelmetadata",
```
Comment (dkalinowski) [inline src/metric_config.hpp:59]: Why do we have those `"ovms_requests_fail_grpc_getmodelmetadata",...` metrics? We never use it in model metric reporter as far I can see.

**Comment by dtrawins** [comment]
with invalid metric name, server fails with incorrect error message:
[servablemanagermodule.cpp:50] ovms::ModelManager::Start() Error: Configuration file not found or cannot open
It should mention invalid metric name instead

File: src/metric_config.hpp [hunk 4]
```diff
@@ -30,73 +32,64 @@ class MetricConfig {
...
+        {"ovms_requests_success",
+            {"ovms_requests_success_grpc_predict",
+                "ovms_requests_success_grpc_getmodelmetadata",
+                "ovms_requests_success_grpc_getmodelstatus",
+                "ovms_requests_success_grpc_modelinfer",
+                "ovms_requests_success_grpc_modelmetadata",
+                "ovms_requests_success_grpc_modelready",
+                "ovms_requests_success_rest_modelinfer",
+                "ovms_requests_success_rest_predict",
+                "ovms_requests_success_rest_get_modelmetadata",
```
[the remark this answers is not in the document]
Comment (rasapala) [reply src/metric_config.hpp:53]: Fixed

File: src/metric_config.hpp (hunk 1 above)
Comment (rasapala) [reply src/metric_config.hpp:54]: Fixed

File: src/metric_config.hpp (hunk 3 above)
Comment (dkalinowski) [reply src/metric_config.hpp:59]: We can add it in the future when we need it. The format of those single metric filters will differ anyway (will not be a single string but json object instead)

**Review by dtrawins — APPROVED** [review]

File: src/metric_config.hpp (hunk 3 above)
Comment (rasapala) [reply src/metric_config.hpp]: Ok

File: src/test/metric_config_test.cpp [hunk 5]
```diff
@@ -121,9 +120,44 @@ TEST_F(MetricsConfigTest, ChangedValues) {
     ASSERT_EQ(metricConfig.metricsEnabled, true);
 
     // ASSERT_EQ(metricConfig.endpointsPath, "/newmetrics");
-    ASSERT_EQ(metricConfig.requestSuccessGrpcPredict, true);
-    ASSERT_EQ(metricConfig.requestFailRestModelReady, true);
-    ASSERT_EQ(metricConfig.requestSuccessGrpcModelReady, false);
+    ASSERT_TRUE(metricConfig.enabledFamiliesList.find("ovms_requests_success") != metricConfig.enabledFamiliesList.end());
```
Comment (dkalinowski) [inline src/test/metric_config_test.cpp:123]: `enabledFamiliesList` could be private field

Comment (dkalinowski) [reply src/test/metric_config_test.cpp:123]: Use `isFamilyEnabled`

Comment (rasapala) [reply src/test/metric_config_test.cpp:123]: Done.

Comment (dkalinowski) [reply src/test/metric_config_test.cpp:123]: All MetricConfig fields and methods are still public

File: src/config.cpp [hunk 6]
```diff
@@ -151,11 +151,13 @@ Config& Config::parse(int argc, char** argv) {
...
             ("metrics_enabled",
-                "Flag indicating if metrics are enabled",
+                "Flag indicating if metrics are enabled.",
                 cxxopts::value<bool>()->default_value("false"),
                 "METRICS")
             ("metrics_list",
-                "List of metrics to be enabled. Comma separated.",
+                R"(List of metrics families to be enabled. If empty the default metrics are enabled:
+                ovms_requests_success, ovms_requests_fail, ovms_request_time_us, ovms_streams, ovms_inference_time_us, ovms_wait_for_infer_req_time_us.
+                Additional families: ovms_infer_req_queue_size, ovms_infer_req_active.)",
```
Comment (dkalinowski) [inline src/config.cpp:160]:
```suggestion
                R"(List of metric families to be enabled. If empty the default metrics are enabled:
                ovms_requests_success, ovms_requests_fail, ovms_request_time_us, ovms_streams, ovms_inference_time_us, ovms_wait_for_infer_req_time_us.
                Additional families: ovms_infer_req_queue_size, ovms_infer_req_active.)",
```

**Review by dkalinowski — APPROVED** [review]

**Pull request merged**